### PR TITLE
chore: check dirs before create RaftEngine store

### LINKED ISSUE #3322: The error message says "Permission denied" but it doesn't say what the file name is
### What type of bug is this?

Other

### What subsystems are affected?

Other

### Minimal reproduce step

start greptime

### What did you expect to see?

n/a

### What did you see instead?

```
Error: 0: Failed to start frontend, at /usr/ports/databases/greptimedb/work/greptimedb-0.7.0-nightly-20240115-102-g29f11d7b7e/src/cmd/src/standalone.rs:400:10
1: Failed to open raft engine backend, at /usr/ports/databases/greptimedb/work/greptimedb-0.7.0-nightly-20240115-102-g29f11d7b7e/src/frontend/src/instance.rs:176:14
2: Failed to perform raft-engine operation, at src/log-store/src/raft_engine/backend.rs:46:43
3: Io(Os { code: 13, kind: PermissionDenied, message: "Permission denied" })
```

### What operating system did you use?

FreeBSD 14.0

### What version of GreptimeDB did you use?

rev. 29f11d7b7e

### Relevant log output and stack trace

_No response_

### PR DESCRIPTION
**Pull request by tisonkun**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This closes #3322.

Also, enrich the IO error with the checked path.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/log-store/src/error.rs [hunk 1]
```diff
@@ -53,6 +53,14 @@ pub enum Error {
         location: Location,
     },
 
+    #[snafu(display("Failed to perform IO on path: {}", path))]
+    Io {
```
Comment (tisonkun) [inline src/log-store/src/error.rs:57]: Doesn't find other proper variants.

File: src/log-store/src/raft_engine/backend.rs [hunk 2]
```diff
@@ -43,11 +43,36 @@ pub struct RaftEngineBackend {
...
+
         let engine = Engine::open(config).context(RaftEngineSnafu)?;
         Ok(Self {
             engine: RwLock::new(engine),
         })
     }
+
+    fn check_config(config: &Config) -> error::Result<()> {
+        fn check_dir(dir: &str) -> error::Result<()> {
+            let metadata = std::fs::metadata(dir).context(IoSnafu {
```
Comment (tisonkun) [inline src/log-store/src/raft_engine/backend.rs:56]: As like -

```
Error: 0: Failed to start frontend, at /Users/tison/Brittani/greptimedb/src/cmd/src/standalone.rs:400:10
1: Failed to open raft engine backend, at /Users/tison/Brittani/greptimedb/src/frontend/src/instance.rs:176:14
2: Failed to perform raft-engine operation, at src/log-store/src/raft_engine/backend.rs:48:43
3: Io(Os { code: 13, kind: PermissionDenied, message: "Permission denied" })
```

cc @yurivict

File: src/log-store/src/raft_engine/backend.rs [hunk 3]
```diff
@@ -43,11 +43,36 @@ pub struct RaftEngineBackend {
...
     pub fn try_open_with_cfg(config: Config) -> error::Result<Self> {
+        Self::check_config(&config)?;
+
         let engine = Engine::open(config).context(RaftEngineSnafu)?;
         Ok(Self {
             engine: RwLock::new(engine),
         })
     }
+
+    fn check_config(config: &Config) -> error::Result<()> {
```
Comment (tisonkun) [inline src/log-store/src/raft_engine/backend.rs:54]: This is somewhat of a monkey patch, but the `std::io::Error` generated from ERRNO doesn't have the path info. RaftEngine's code is complex to enrich the info, as well as its IO error variant doesn't have a message field. Let's fix this case specially, although I may prefer a global utility for enriching IO errors.

**Review by waynexia — APPROVED** [review]
LGTM

**Review by MichaelScofield — APPROVED** [review]

File: src/log-store/src/raft_engine/backend.rs (hunk 3 above)
Comment (waynexia) [reply src/log-store/src/raft_engine/backend.rs]: Most of fs operations are through OpenDAL which has a better error behavior. But unfortunately raft-engine is an exception

**Pull request merged**